packages/instantsearch.js/src/lib/utils/hydrateSearchClient.ts

```diff
@@ -34,16 +34,24 @@ export function hydrateSearchClient(
     return;
   }
 
-  const cachedRequest = Object.keys(results).map((key) =>
-    results[key].results.map((result) => ({
-      indexName: result.index,
+  const cachedRequest = Object.keys(results).map((key) => {
+    const { state, requestParams, results: serverResults } = results[key];
+    return serverResults.map((result) => ({
+      indexName: result.index || state.index,
```

Comment:
```suggestion
      indexName: state.index || result.index,
```

Reply:
this should be good if we ever add a feature in the backend that redirects indices (I believe I've heard of it, not sure if it ever came to fruition)